File: services/backend/src/domain/validation/tax_id.py
import re


def _char_value(c: str) -> int:
    """Numeric value of a CNPJ character: ASCII code − 48 (0→0…9→9, A→17…Z→42)."""
    return ord(c.upper()) - 48


def _validate_cpf(raw: str) -> None:
    clean = re.sub(r'\D', '', raw)
    if len(clean) != 11:
        raise ValueError(f"CPF must have 11 digits (got {len(clean)})")
    if len(set(clean)) == 1:
        raise ValueError("CPF must not consist of a single repeated digit")

    s1 = sum(int(clean[i]) * (10 - i) for i in range(9))
    r1 = s1 % 11
    d1 = 0 if r1 < 2 else 11 - r1
    if int(clean[9]) != d1:
        raise ValueError("Invalid CPF: first check digit mismatch")

    s2 = sum(int(clean[i]) * (11 - i) for i in range(10))
    r2 = s2 % 11
    d2 = 0 if r2 < 2 else 11 - r2
    if int(clean[10]) != d2:
        raise ValueError("Invalid CPF: second check digit mismatch")


def _validate_cnpj(raw: str) -> None:
    """
    Validate a Brazilian CNPJ (alphanumeric format, IN RFB 2.229/2024).

    Format: 14 characters total.
      - Positions 1–8  : [A-Z0-9]  (raiz)
      - Positions 9–12 : [A-Z0-9]  (filial / order)
      - Positions 13–14: [0-9]     (check digits, modulo-11)

    Character values: ASCII code − 48 (0→0…9→9, A→17…Z→42).
    """
    clean = re.sub(r'[.\-/]', '', raw.upper())

    if len(clean) != 14:
        raise ValueError(f"CNPJ must have 14 characters (got {len(clean)})")

    if not re.match(r'^[A-Z0-9]{12}\d{2}$', clean):
        raise ValueError("CNPJ: positions 13–14 must be digits")

    if len(set(clean)) == 1:
        raise ValueError("CNPJ must not consist of a single repeated character")

    # First check digit — weights over positions 1–12
    w1  = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    s1  = sum(_char_value(clean[i]) * w1[i] for i in range(12))
    r1  = s1 % 11
    d1  = 0 if r1 < 2 else 11 - r1
    if int(clean[12]) != d1:
        raise ValueError("Invalid CNPJ: first check digit mismatch")

    # Second check digit — weights over positions 1–13
    w2  = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    s2  = sum(_char_value(clean[i]) * w2[i] for i in range(13))
    r2  = s2 % 11
    d2  = 0 if r2 < 2 else 11 - r2
    if int(clean[13]) != d2:
        raise ValueError("Invalid CNPJ: second check digit mismatch")


# ---------------------------------------------------------------------------
# Country dispatch


def validate_tax_id(country: str, raw: str) -> None:
    """Validate a tax ID for the given ISO-3166-1 alpha-2 country code.
    Raises ValueError with a human-readable message on failure.
    Unknown countries are silently accepted (future extensibility).
    For Brazil: 11-digit raw → CPF; 14-character raw → CNPJ.
    """
    if country.upper() == "BR":
        digits_only = re.sub(r'\D', '', raw)
        if len(digits_only) == 11:
            _validate_cpf(raw)
        else:
            _validate_cnpj(raw)

File: services/web/src/shared/taxId.ts
// Tax ID utilities — alphanumeric CNPJ support (Brazil, IN RFB 2.229/2024)
//
// New CNPJ format (effective July 2026): positions 1–12 may be [A-Z0-9],
// only positions 13–14 (check digits) remain numeric.
// Character values: ASCII code − 48  (0→0…9→9, A→17…Z→42).

function charValue(c: string): number {
  return c.toUpperCase().charCodeAt(0) - 48            // ASCII − 48 for all chars
}

// ---------------------------------------------------------------------------
// CNPJ

export function stripCNPJ(v: string): string {
  return v.toUpperCase().replace(/[.\-/]/g, '').replace(/[^A-Z0-9]/g, '')
}

export function maskCNPJ(raw: string): string {
  // Strip separators, uppercase, then build clean string:
  // positions 0–11 accept A-Z0-9, positions 12–13 (check digits) accept digits only.
  const src = raw.toUpperCase().replace(/[.\-/]/g, '')
  let clean = ''
  for (const ch of src) {
    if (clean.length >= 14) break
    if (clean.length < 12) {
      if (/[A-Z0-9]/.test(ch)) clean += ch
    } else {
      if (/[0-9]/.test(ch)) clean += ch
    }
  }
  const n = clean.length
  if (n <= 2)  return clean
  if (n <= 5)  return `${clean.slice(0, 2)}.${clean.slice(2)}`
  if (n <= 8)  return `${clean.slice(0, 2)}.${clean.slice(2, 5)}.${clean.slice(5)}`
  if (n <= 12) return `${clean.slice(0, 2)}.${clean.slice(2, 5)}.${clean.slice(5, 8)}/${clean.slice(8)}`
  return `${clean.slice(0, 2)}.${clean.slice(2, 5)}.${clean.slice(5, 8)}/${clean.slice(8, 12)}-${clean.slice(12)}`
}

export function validateCNPJ(raw: string): boolean {
  const clean = stripCNPJ(raw)
  if (clean.length !== 14) return false
  if (new Set(clean).size === 1) return false          // all same char
  if (!/^\d{2}$/.test(clean.slice(12))) return false  // only check digits must be numeric

  const w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
  const s1  = w1.reduce((acc, w, i) => acc + charValue(clean[i]) * w, 0)
  const r1  = s1 % 11
  const d1  = r1 < 2 ? 0 : 11 - r1
  if (parseInt(clean[12], 10) !== d1) return false

  const w2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
  const s2  = w2.reduce((acc, w, i) => acc + charValue(clean[i]) * w, 0)
  const r2  = s2 % 11
  const d2  = r2 < 2 ? 0 : 11 - r2
  return parseInt(clean[13], 10) === d2
}

export function cnpjError(raw: string): string | null {
  const clean = stripCNPJ(raw)
  if (clean.length === 0) return null
  if (clean.length < 14)  return 'CNPJ must have 14 characters'
  if (!validateCNPJ(raw))  return 'Invalid CNPJ — check digit mismatch'
  return null
}

// ---------------------------------------------------------------------------
// CPF

export function stripCPF(v: string): string {
  return v.replace(/\D/g, '')
}

export function maskCPF(raw: string): string {
  const digits = raw.replace(/\D/g, '').slice(0, 11)
  const n = digits.length
  if (n <= 3)  return digits
  if (n <= 6)  return `${digits.slice(0, 3)}.${digits.slice(3)}`
  if (n <= 9)  return `${digits.slice(0, 3)}.${digits.slice(3, 6)}.${digits.slice(6)}`
  return `${digits.slice(0, 3)}.${digits.slice(3, 6)}.${digits.slice(6, 9)}-${digits.slice(9)}`
}

export function validateCPF(raw: string): boolean {
  const clean = raw.replace(/\D/g, '')
  if (clean.length !== 11) return false
  if (new Set(clean).size === 1) return false

  const s1 = Array.from({ length: 9 }, (_, i) => parseInt(clean[i], 10) * (10 - i)).reduce((a, b) => a + b, 0)
  const r1  = s1 % 11
  const d1  = r1 < 2 ? 0 : 11 - r1
  if (parseInt(clean[9], 10) !== d1) return false

  const s2 = Array.from({ length: 10 }, (_, i) => parseInt(clean[i], 10) * (11 - i)).reduce((a, b) => a + b, 0)
  const r2  = s2 % 11
  const d2  = r2 < 2 ? 0 : 11 - r2
  return parseInt(clean[10], 10) === d2
}

export function cpfError(raw: string): string | null {
  const clean = raw.replace(/\D/g, '')
  if (clean.length === 0) return null
  if (clean.length < 11)  return null
  if (!validateCPF(raw))  return 'CPF inválido'
  return null
}

// ---------------------------------------------------------------------------
// ZIP / CEP

export function maskZip(country: string, raw: string): string {
  if (country === 'BR') {
    const digits = raw.replace(/\D/g, '').slice(0, 8)
    if (digits.length <= 5) return digits
    return `${digits.slice(0, 5)}-${digits.slice(5)}`
  }
  return raw.replace(/\D/g, '')
}

export function zipPlaceholder(country: string): string {
  if (country === 'BR') return '00000-000'
  return 'ZIP Code'
}

// ---------------------------------------------------------------------------
// Country dispatch
// taxIdType: 'cpf' forces CPF mode for Brazilian individuals; default is CNPJ

export function maskTaxId(country: string, raw: string, taxIdType?: 'cpf' | 'cnpj'): string {
  if (country === 'BR') {
    if (taxIdType === 'cpf') return maskCPF(raw)
    return maskCNPJ(raw)
  }
  return raw
}

export function validateTaxId(country: string, raw: string, taxIdType?: 'cpf' | 'cnpj'): boolean {
  if (country === 'BR') {
    if (taxIdType === 'cpf') return validateCPF(raw)
    return validateCNPJ(raw)
  }
  return true  // unknown country: skip client-side validation
}

export function taxIdError(country: string, raw: string, taxIdType?: 'cpf' | 'cnpj'): string | null {
  if (country === 'BR') {
    if (taxIdType === 'cpf') return cpfError(raw)
    return cnpjError(raw)
  }
  return null
}

export function taxIdPlaceholder(country: string, taxIdType?: 'cpf' | 'cnpj'): string {
  if (country === 'BR') {
    if (taxIdType === 'cpf') return '000.000.000-00'
    return 'XX.XXX.XXX/XXXX-XX'
  }
  return 'Tax ID'
}

export function taxIdLabel(country: string, taxIdType?: 'cpf' | 'cnpj'): string {
  if (country === 'BR') {
    if (taxIdType === 'cpf') return 'CPF'
    return 'CNPJ'
  }
  return 'Tax ID'
}
